### PR TITLE
Step 8 – kube-apiserver: split up config creation into generic and non-generic part

### PR DESCRIPTION
**Pull request by sttts**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Next step after https://github.com/kubernetes/kubernetes/pull/124621.

This PR splits up the kube-apiserver config construction into the generic controlplane part and kube-specific part. In other words, it adds the controlplane layer in `kube-apiserver Config, controlplane Config, generic apiserver Config`.

NO FUNCTIONAL CHANGE.

Part of #124530.

#### Which issue(s) this PR fixes:

Towards generic controlplanes kubernetes/enhancements#4080.

#### Special notes for your reviewer:

Look careful that this does not change any behaviour.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by sttts** [comment]
Integration might be a flake. Works locally.

/retest

**Comment by mjudeikis** [comment]
/lgtm 
easy enough to move on :)

**Comment by mjudeikis** [comment]
/lgtm-remove
need rebase first

**Comment by mjudeikis** [comment]
/remove-lgtm
needs rebase first

**Comment by sttts** [comment]
Infra.

/retest

File: pkg/controlplane/apiserver/config.go [hunk 1]
```diff
@@ -254,3 +262,131 @@ func BuildAuthorizer(ctx context.Context, s controlplaneapiserver.CompletedOptio
...
+	versionedInformers clientgoinformers.SharedInformerFactory,
+	storageFactory *serverstorage.DefaultStorageFactory,
+	serviceResolver aggregatorapiserver.ServiceResolver,
+	additionalInitializers []admission.PluginInitializer,
+) (
+	*Config,
+	[]admission.PluginInitializer,
+	error,
+) {
+	proxyTransport := CreateProxyTransport()
```
Comment (mjudeikis) [inline pkg/controlplane/apiserver/config.go:280]: Just wondering, should be part of k-apiserver, not control-plane? as there are no nodes in generic controlplane

File: pkg/controlplane/apiserver/config.go [hunk 2]
```diff
@@ -254,3 +262,131 @@ func BuildAuthorizer(ctx context.Context, s controlplaneapiserver.CompletedOptio
...
+			ServiceAccountMaxExpiration: opts.ServiceAccountTokenMaxExpiration,
+			ExtendExpiration:            opts.Authentication.ServiceAccounts.ExtendExpiration,
+
+			VersionedInformers: versionedInformers,
+		},
+	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.UnknownVersionInteroperabilityProxy) {
+		var err error
+		config.PeerEndpointLeaseReconciler, err = CreatePeerEndpointLeaseReconciler(*genericConfig, storageFactory)
```
Comment (mjudeikis) [inline pkg/controlplane/apiserver/config.go:305]: same this? Do we see this needed on Generic CP?

Comment (sttts) [reply pkg/controlplane/apiserver/config.go:305]: Peer endpoints are generic. It‘s cross-apiserver communication.

**Comment by mjudeikis** [comment]
This changes a bit how kube-apiserver is instanciated, but sampes in last PR should cover this ;) 
/lgtm

File: pkg/controlplane/apiserver/config.go [hunk 3]
```diff
@@ -254,3 +262,131 @@ func BuildAuthorizer(ctx context.Context, s controlplaneapiserver.CompletedOptio
...
+	config.ServiceAccountJWKSURI = opts.Authentication.ServiceAccounts.JWKSURI
+	config.ServiceAccountPublicKeys = pubKeys
+
+	return config, genericInitializers, nil
+}
+
+// CreateProxyTransport creates the dialer infrastructure to connect to the nodes.
+func CreateProxyTransport() *http.Transport {
+	var proxyDialerFn utilnet.DialFunc
+	// Proxying to pods and services is IP-based... don't expect to be able to verify the hostname
```
Comment (luxas) [inline pkg/controlplane/apiserver/config.go:385]: Do we have an issue to track this; if anything could be done e.g. in the generic API server case, with slightly altered, non-kube assumptions?

**Review by luxas — COMMENTED** [review]
Getting closer to a full split 💯🎉

**Pull request merged**